services/data/src/hooks/useDataQuery.ts

```diff
@@ -1,29 +1,62 @@
-import { useState, useContext, useEffect } from 'react'
+import { useState, useContext, useEffect, useCallback } from 'react'
 import { DataContext } from '../components/DataContext'
-import { QueryState, QueryMap } from '../types/Query'
+import {
+    QueryState,
+    QueryMap,
+    RefetchCallback,
+    QueryRenderInput,
+} from '../types/Query'
+import { ContextType } from '../types/Context'
 
-export const useDataQuery = (query: QueryMap): QueryState => {
+const reduceResponses = (responses: any[], names: string[]) =>
+    responses.reduce((out, response, idx) => {
+        out[names[idx]] = response
+    }, {})
+
+const fetchData = (
+    context: ContextType,
+    query: QueryMap,
+    signal: AbortSignal
+) => {
     const names = Object.keys(query)
     const requests = names.map(name => query[name])
 
-    const context = useContext(DataContext)
+    const requestPromises = requests.map(q =>
+        context.fetch(q, {
+            signal: signal,
+        })
+    )
+
+    return Promise.all(requestPromises).then(responses =>
+        reduceResponses(responses, names)
+    )
+}
 
+export const useDataQuery = (query: QueryMap): QueryRenderInput => {
+    const context = useContext(DataContext)
     const [state, setState] = useState<QueryState>({ loading: true })
+    const [refetchCount, setRefetchCount] = useState(0)
+    const refetch: RefetchCallback = useCallback(
+        () => setRefetchCount(count => count + 1),
+        []
+    )
 
     useEffect(() => {
-        Promise.all(requests.map(q => context.fetch(q)))
-            .then(responses =>
-                responses.reduce(
-                    (out, response, idx) => ({
-                        ...out,
-                        [names[idx]]: response,
-                    }),
-                    []
-                )
-            )
-            .then(data => setState({ loading: false, data }))
-            .catch(error => setState({ loading: false, error }))
-    }, [])
-
-    return state
+        const controller = new AbortController()
```

Comment:
IE11 lacks support for the `AbortController`. From DHIS 2 version 2.33 we will drop support for IE 11, so most apps using this hook won't suffer from this problem. However, there may be some exceptions:
- Apps that transitioned to feature toggling prior to 2.33: if at a later stage a developer decides to introduce the DataQuery component, this could cause problems (they can probably be avoided somehow)
- The new HeaderBar component: I believe this uses the DataQuery too, and it is being used in a lot of v2.32 apps. So this could be another potential source of problems.

My preferred solution would be to do the following:
- Keep the code as it is
- Add a section to the README (preferably fairly prominently displayed) mentioning:
   - That this lib relies on the `Fetch`, `AbortController` and `AbortSignal` APIs which are unsupported in IE11
   - If your app requires IE11 support, add a polyfill for all these APIs
   - Include some instructions on how to do so, i.e. linking to [whatwg-fetch](https://github.com/github/fetch) and [abortcontroller-polyfill](https://github.com/mo/abortcontroller-polyfill)

This each app can handle things as it sees fit.